### PR TITLE
perf(dwarf): share one DWARFInfo across metadata + snapshot passes

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

Thread one open `DWARFInfo` through all three DWARF extraction passes so a DWARF-only dump parses each DIE once instead of once per pass — byte-for-byte identical output.

## Motivation

DWARF-only dumps ran the basic-metadata, advanced-metadata, and snapshot-builder passes as three independent ELF opens, each building its own `DWARFInfo` with a cold DIE cache. On a large C++ surface the same `std::`/template subtrees get re-parsed once per pass — the **F5b** finding from the pvxs validation (`validation/pvxs-abi-validation-2026-07.md`), where the snapshot pass alone was ~30 s on `libpvxs`, dominated by DIE re-parse. This is the "single-pass merge" step of that follow-up.

The naive "skip already-seen types" fix was measured **wrong** in F5b (dropped 195 types — later CUs contribute nested members the first didn't). This change instead shares the parse handle, which changes *no* output: pyelftools caches parsed DIEs within a CU and caches CU objects across `iter_CUs()`, so a later pass over the same `DWARFInfo` is served from the cache the earlier passes warmed.

## Changes

- **`dwarf_unified`** — add `DwarfSession` (one open ELF + one `DWARFInfo`), `open_dwarf_session`, and `parse_dwarf_from_session`. `parse_dwarf` keeps its exact 2-tuple contract and closes by default; an internal `_session_out` sink hands the still-open session back for reuse.
- **`dwarf_snapshot`** — `build_snapshot_from_dwarf` / `extract` accept an optional `session` and walk its `DWARFInfo` instead of re-opening the ELF. The built snapshot holds extracted model objects (no live DIE refs), so it stays valid after the session closes.
- **`dumper._dump_elf`** — opens one session via `_resolve_debug_metadata(..., _session_out=…)`, reuses it for the snapshot build, and closes it in a `finally` on every exit path (snapshot / symbol-only / header).
- **`dumper_debug`** (new) — `_resolve_debug_metadata` + `_is_kernel_binary` moved out of `dumper.py` (which was at the 2000-line hard cap) and re-exported so `dumper._is_kernel_binary` / `dumper._resolve_debug_metadata` remain valid bare-name calls and test patch targets.
- Tests: session byte-identical guarantee, reuse-without-reopen, post-close serializability, and `open_dwarf_session` edge cases.

### Validation (A/B on real compiled binaries)

On a 9-CU C++ fixture (repeated template/`std::` types across CUs, like pvxs):

| | before (independent opens) | after (shared session) |
|---|---|---|
| snapshot build | ~0.58 s | **~0.07 s** |

Snapshot, basic-metadata, and advanced-metadata JSON are **byte-for-byte identical** across (a) pre-refactor baseline vs refactored, and (b) no-session vs session paths. End-to-end `compare`/`dump` verified; no fd leak across repeated dumps. No CLI or public-API surface change.

## Checklist

- [x] Tests added / updated for new behaviour
- [x] `CHANGELOG.md` updated (under `[Unreleased]`)
- [x] Docs updated if user-visible behaviour changed (n/a — internal only; module docstrings updated)
- [x] No new `mypy` or `ruff` errors introduced (`mypy abicheck/` clean, `ruff check abicheck/ tests/` clean, AI-readiness 0 errors)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01LqUUSp8wT2FSqPQodvwpeh

---
_Generated by [Claude Code](https://claude.ai/code/session_01LqUUSp8wT2FSqPQodvwpeh)_

## Summary by CodeRabbit

* **Performance**
  * Improved DWARF-only dump performance by reusing parsed debug information across multiple extraction passes.
  * Reduced redundant debug-information processing while keeping results byte-for-byte identical.
* **Reliability**
  * More robust debug metadata resolution, with safer handling when requested formats or debug sections are missing.
* **Tests**
  * Added new session-sharing and performance integration coverage, including timing comparisons, scaling guards, and byte-identical snapshot verification.

### CODE REVIEW

File: tests/test_dwarf_unified.py [hunk 1]
```diff
@@ -234,3 +237,145 @@ def counting_open(path, mode="r", **kwargs):  # type: ignore[override]
...
+                reopens.append(str(path))
+            return original_open(path, mode, **kwargs)
+
+        try:
+            with patch("builtins.open", side_effect=counting_open):
+                build_snapshot_from_dwarf(so, elf_meta, meta, adv, session=sess)
+        finally:
+            sess.close()
+
+        assert reopens == [], f"snapshot re-opened the ELF despite a session: {reopens}"
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_dwarf_unified.py:448]: Intentionally not adding `@pytest.mark.integration` here. The sibling classes in this same file that compile with gcc/g++ (`TestUnifiedEqualsSepaRate`, `TestShims`, `TestSingleOpen`) are deliberately unmarked and run in the default/coverage lane — this repo's `integration` marker is scoped to **castxml**-requiring tests (per `tests/CLAUDE.md`), and these only need g++ with a `_require_tool` guard + Linux skip. Marking `TestDwarfSession` would (a) diverge from that convention and (b) pull these tests out of the Linux coverage lane, which would *lower* patch coverage on exactly the new session code they exist to cover. Keeping them consistent with the existing g++ tests instead.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
## Closed a gap: no perf/scalability regression guard for the DwarfSession win

The DWARFInfo-sharing optimization was validated correctly (byte-identical output, unit tests) but the *performance win itself* had zero automated protection — only a one-off manual A/B measurement in a local scratchpad, never committed. Nothing in CI would have caught a future regression back to independent per-pass ELF opens, and the win was never checked at scale (varying CU count).

Added in `2b93c7c` — `tests/test_perf_dwarf_session_scaling.py` (new, `integration`):

- **`test_session_reuse_faster_than_independent_opens`** — compiles a real multi-CU C++ binary (14 CUs, each `#include`-ing a shared header with a repeated template/`std::` type — reproducing the pvxs pathology), then runs the legacy independent-open path and the current shared-session path back-to-back in-process. Asserts the session path is reliably faster (threshold `0.85x`, conservative — measured ratio is consistently ~0.5x locally across repeated runs, so there's real margin against CI noise). This is a deterministic regression guard for the actual mechanism: if `build_snapshot_from_dwarf` ever stops honoring `session=`, this fails.
- **`test_dwarf_only_dump_scaling_with_cu_count_stays_subquadratic`** — times the production `dumper.dump(..., dwarf_only=True)` path at 8 vs 32 CUs, exponent-gated `<1.9` (matching the convention in `test_perf_dump_scaling.py`). Guards against a hidden O(n²) as CU count grows — an axis the existing single-CU scaling test doesn't vary.

Deliberately **not** wired into `.github/workflows/performance.yml`'s path filters: that lane's `scripts/benchmark_scaling.py` builds `AbiSnapshot` objects directly and never invokes a real compiler (by design — see `test_benchmark_scaling.py`), so it structurally cannot exercise DWARF parsing. The `integration` marker is the correct wiring instead — `ci.yml`'s `integration-tests` job runs `-m integration` unconditionally on every push/PR (all three OS lanes on Linux via castxml+gcc+g++), no path filter needed.

Verified locally: both new tests pass, and the timing test's margin is well clear of its threshold (~0.48–0.53 measured ratio vs 0.85 gate) across 4 repeated runs — not flake-prone.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/dwarf_unified.py [hunk 2]
```diff
@@ -56,59 +69,146 @@
...
 
-            meta = DwarfMetadata(has_dwarf=True)
-            adv = AdvancedDwarfMetadata(has_dwarf=True)
-            adv.target_arch = _normalize_arch(elf)
+    # Per-binary type-resolution cache: (cu_offset, die_offset) → (type_name, byte_size).
+    # DIE offsets are only unique within one ELF file — do not share across binaries.
+    type_cache: dict[tuple[int, int], tuple[str, int]] = {}
 
-            dwarf = elf.get_dwarf_info()  # type: ignore[no-untyped-call]
+    for CU in session.dwarf.iter_CUs():
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/dwarf_unified.py:168]: Fixed in commit 61fc635. `parse_dwarf` now wraps `parse_dwarf_from_session(session)` in a `try/except Exception`: on failure it closes the session and returns empty `(DwarfMetadata(), AdvancedDwarfMetadata())` instead of letting the exception (and the leaked handle) propagate. Covered by `test_parse_dwarf_survives_cu_iteration_failure` in `tests/test_dwarf_unified.py`, which monkeypatches `parse_dwarf_from_session` to raise and asserts the session is closed (not appended to the caller's out-list) and empty metadata comes back.

---
_Generated by [Claude Code](https://claude.ai/code)_
